lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client.timestampleases;
+
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.palantir.atlasdb.timelock.api.ConjureLockToken;
+import com.palantir.atlasdb.timelock.api.LeaseGuarantee;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import com.palantir.atlasdb.timelock.api.RequestId;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseName;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseRequests;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseResponse;
+import com.palantir.atlasdb.timelock.api.TimestampLeaseResponses;
+import com.palantir.atlasdb.timelock.api.TimestampLeasesRequest;
+import com.palantir.atlasdb.timelock.api.TimestampLeasesResponse;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.lock.client.InternalMultiClientConjureTimelockService;
+import com.palantir.lock.client.LeasedLockToken;
+import com.palantir.lock.client.TimeLockUnlocker;
+import com.palantir.lock.v2.ImmutableTimestampLeaseResult;
+import com.palantir.lock.v2.ImmutableTimestampLeaseResults;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.lock.v2.TimestampLeaseResult;
+import com.palantir.lock.v2.TimestampLeaseResults;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.tritium.ids.UniqueIds;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import org.immutables.value.Value;
+
+final class TimestampLeaseAcquirerImpl implements TimestampLeaseAcquirer {
+    private static final SafeLogger log = SafeLoggerFactory.get(TimestampLeaseAcquirerImpl.class);
+
+    private final Retryer<RequestAndResponse> retryer = RetryerBuilder.<RequestAndResponse>newBuilder()
+            .retryIfResult(requestAndResponse -> !wasRequestFullyFulfilled(requestAndResponse))
+            .retryIfException()
+            .withStopStrategy(StopStrategies.stopAfterAttempt(3))
+            .build();
+
+    private final Namespace namespace;
+    private final InternalMultiClientConjureTimelockService service;
+    private final TimeLockUnlocker unlocker;
```

Comment:
We actually have a nice implementation for this https://github.com/palantir/atlasdb/blob/0.1171.0/lock-api/src/main/java/com/palantir/lock/client/AsyncTimeLockUnlocker.java.

However, using that one will be awkward. It is defined at the level of `TimelockClient` and uses `TimelockService` as a dependency. Sort of the final step / things given to TMs.

I have a couple of options:
- Do ugly up-down-up wiring to get an instance of this 
- Implement my own class with some sort of executor spun up
- Copy the class and change the dependency from TimelockService to LockLeaseService or a generic function interface (that could even make it so I can reuse in the original if I just refactor)

The first is just too much tech debt.

The second one would be costly for internal skiing product (one executor per client), and error-prone.

The third one seems most realistic to me (with the duplication and its own disruptor, which can eventually have internal skiing project pass something to).

Thoughts?

Reply:
Well: the ACTUAL way this should work is that LockToken is closeable :) SO then you don't need to write this up into 50 places, instead when a lock token is given out, it KNOWs how to unlock itself.